File: pageindicatorview/src/main/java/com/rd/draw/controller/DrawController.java
package com.rd.draw.controller;

import android.graphics.Canvas;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.MotionEvent;
import com.rd.animation.data.Value;
import com.rd.animation.type.AnimationType;
import com.rd.draw.data.Indicator;
import com.rd.draw.drawer.Drawer;
import com.rd.utils.CoordinatesUtils;

public class DrawController {

	private Value value;
	private Drawer drawer;
	private Indicator indicator;
	private ClickListener listener;

	public interface ClickListener {

		void onIndicatorClicked(int position);
	}

	public DrawController(@NonNull Indicator indicator) {
		this.indicator = indicator;
		this.drawer = new Drawer(indicator);
	}

	public void updateValue(@Nullable Value value) {
        this.value = value;
    }

	public void setClickListener(@Nullable ClickListener listener) {
		this.listener = listener;
	}

	public void touch(@Nullable MotionEvent event) {
		if (event == null) {
			return;
		}

		switch (event.getAction()) {
			case MotionEvent.ACTION_UP:
				onIndicatorTouched(event.getX(), event.getY());
				break;
			default:
		}
	}

	private void onIndicatorTouched(float x, float y) {
		if (listener != null) {
			int position = CoordinatesUtils.getPosition(indicator, x, y);
			if (position >= 0) {
				listener.onIndicatorClicked(position);
			}
		}
	}

	public void draw(@NonNull Canvas canvas) {
        int count = indicator.getCount();

        for (int position = 0; position < count; position++) {
            int coordinateX = CoordinatesUtils.getXCoordinate(indicator, position);
            int coordinateY = CoordinatesUtils.getYCoordinate(indicator, position);
            drawIndicator(canvas, position, coordinateX, coordinateY);
        }
    }

    private void drawIndicator(
            @NonNull Canvas canvas,
            int position,
            int coordinateX,
            int coordinateY) {

        boolean interactiveAnimation = indicator.isInteractiveAnimation();
        int selectedPosition = indicator.getSelectedPosition();
        int selectingPosition = indicator.getSelectingPosition();
        int lastSelectedPosition = indicator.getLastSelectedPosition();

        boolean selectedItem = !interactiveAnimation && (position == selectedPosition || position == lastSelectedPosition);
        boolean selectingItem = interactiveAnimation && (position == selectedPosition || position == selectingPosition);
        boolean isSelectedItem = selectedItem | selectingItem;
        drawer.setup(position, coordinateX, coordinateY);

        if (value != null && isSelectedItem) {
            drawWithAnimation(canvas);
        } else {
            drawer.drawBasic(canvas, isSelectedItem);
        }
    }

    private void drawWithAnimation(@NonNull Canvas canvas) {
        AnimationType animationType = indicator.getAnimationType();
        switch (animationType) {
            case NONE:
                drawer.drawBasic(canvas, true);
                break;

            case COLOR:
                drawer.drawColor(canvas, value);
                break;

            case SCALE:
                drawer.drawScale(canvas, value);
                break;

            case WORM:
                drawer.drawWorm(canvas, value);
                break;

            case SLIDE:
                drawer.drawSlide(canvas, value);
                break;

            case FILL:
                drawer.drawFill(canvas, value);
                break;

            case THIN_WORM:
                drawer.drawThinWorm(canvas, value);
                break;

            case DROP:
                drawer.drawDrop(canvas, value);
                break;

            case SWAP:
                drawer.drawSwap(canvas, value);
                break;

            case SCALE_DOWN:
                drawer.drawScaleDown(canvas, value);
                break;
        }
    }
}

File: pageindicatorview/src/main/java/com/rd/draw/controller/DrawController.kt
package com.rd.draw.controller

import android.graphics.Canvas
import android.view.MotionEvent
import com.rd.animation.data.Value
import com.rd.animation.type.AnimationType
import com.rd.draw.data.Indicator
import com.rd.draw.drawer.Drawer
import com.rd.utils.CoordinatesUtils

class DrawController(private val indicator: Indicator) {

	private var value: Value? = null
	private val drawer = Drawer(indicator)
	private var listener: ClickListener? = null

	interface ClickListener {

		fun onIndicatorClicked(position: Int)
	}

	fun updateValue(value: Value?) {
		this.value = value
	}

	fun setClickListener(listener: ClickListener?) {
		this.listener = listener
	}

	fun touch(event: MotionEvent?) {
		if (event == null) {
			return
		}

		when (event.action) {
			MotionEvent.ACTION_UP -> onIndicatorTouched(event.x, event.y)
		}
	}

	private fun onIndicatorTouched(x: Float, y: Float) {
		if (listener != null) {
			val position = CoordinatesUtils.getPosition(indicator, x, y)
			if (position >= 0) {
				listener!!.onIndicatorClicked(position)
			}
		}
	}

	fun draw(canvas: Canvas) {
		val count = indicator.count

		for (position in 0 until count) {
			val coordinateX = CoordinatesUtils.getXCoordinate(indicator, position)
			val coordinateY = CoordinatesUtils.getYCoordinate(indicator, position)
			drawIndicator(canvas, position, coordinateX, coordinateY)
		}
	}

	private fun drawIndicator(
			canvas: Canvas,
			position: Int,
			coordinateX: Int,
			coordinateY: Int) {

		val interactiveAnimation = indicator.isInteractiveAnimation
		val selectedPosition = indicator.selectedPosition
		val selectingPosition = indicator.selectingPosition
		val lastSelectedPosition = indicator.lastSelectedPosition

		val selectedItem = !interactiveAnimation && (position == selectedPosition || position == lastSelectedPosition)
		val selectingItem = interactiveAnimation && (position == selectedPosition || position == selectingPosition)
		val isSelectedItem = selectedItem or selectingItem
		drawer.setup(position, coordinateX, coordinateY)

		if (value != null && isSelectedItem) {
			drawWithAnimation(canvas)
		} else {
			drawer.drawBasic(canvas, isSelectedItem)
		}
	}

	private fun drawWithAnimation(canvas: Canvas) {
		if (value == null || value !is Value) return
		val animationValue = value as Value

		when (indicator.animationType) {
			AnimationType.NONE -> drawer.drawBasic(canvas, true)

			AnimationType.COLOR -> drawer.drawColor(canvas, animationValue)

			AnimationType.SCALE -> drawer.drawScale(canvas, animationValue)

			AnimationType.WORM -> drawer.drawWorm(canvas, animationValue)

			AnimationType.SLIDE -> drawer.drawSlide(canvas, animationValue)

			AnimationType.FILL -> drawer.drawFill(canvas, animationValue)

			AnimationType.THIN_WORM -> drawer.drawThinWorm(canvas, animationValue)

			AnimationType.DROP -> drawer.drawDrop(canvas, animationValue)

			AnimationType.SWAP -> drawer.drawSwap(canvas, animationValue)

			AnimationType.SCALE_DOWN -> drawer.drawScaleDown(canvas, animationValue)
		}
	}
}
